src/drivers/uart/src/amdmmio.rs

```diff
@@ -0,0 +1,140 @@
+/*
+ * This file is part of the oreboot project.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 2 of the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+use core::ops;
+use model::*;
+
+use register::mmio::{ReadOnly, ReadWrite};
+use register::register_bitfields;
+
+const RETRY_COUNT: u32 = 100_000;
+const COM1: usize = 0xfedc9000;
+
+// We fill out as little of this as possible.
+// We're firmware and should never plan to use it
+// with all its features.
+#[repr(C)]
+pub struct RegisterBlock {
+    // pathetically, this is also the DLL.
+    d: ReadWrite<u32, D::Register>, /* data register */
+    // This is the IER when we are not in DLAB mode.
+    // We never intend to set inerrupts, so we just set
+    // it up as thing we write to.
+    dlm: ReadWrite<u32, D::Register>, /* data register */
+    _8: u32,
+    lcr: ReadWrite<u32, LCR::Register>,
+    _10: u32,
+    stat: ReadOnly<u32, STAT::Register>, /* status */
+}
+
+pub struct AMDMMIO {
+    base: usize,
+    // TODO: implement baudrate
+    //baudrate: u32,
+}
+
+impl ops::Deref for AMDMMIO {
+    type Target = RegisterBlock;
+    fn deref(&self) -> &Self::Target {
+        unsafe { &*self.ptr() }
+    }
+}
+
+register_bitfields! {
+    u32,
+    D [
+        Data OFFSET(0) NUMBITS(8) []
+    ],
+    STAT[
+        DR OFFSET(0) NUMBITS(1) [],
+        THRE OFFSET(6) NUMBITS(1) [],
+        TDRE OFFSET(5) NUMBITS(1) []
+    ],
+    FCR[
+        FIFOENABLE OFFSET(0) NUMBITS(1) []
+    ],
+    LCR[
+        BITSPARITY OFFSET(1) NUMBITS(3) [
+            EIGHTN1 = 3
+        ],
+        
+        DLAB OFFSET(7) NUMBITS(1)[]
+    ]
+}
+
+impl AMDMMIO {
+    pub fn new(base: usize) -> AMDMMIO {
+        AMDMMIO { base: base }
+    }
+
+    pub fn com1() -> AMDMMIO {
+        AMDMMIO { base: COM1 }
+    }
+
+    /// Returns a pointer to the register block
+    fn ptr(&self) -> *const RegisterBlock {
+        self.base as *const _
+    }
+}
+
+impl Driver for AMDMMIO {
+    fn init(&mut self) -> Result<()> {
+        // This does not work.
+//         if false {
+            self.lcr.modify(LCR::BITSPARITY::EIGHTN1);
+            self.lcr.modify(LCR::DLAB::SET);
+            self.d.set(1);
+            self.dlm.set(0);
+            self.lcr.modify(LCR::DLAB::CLEAR);
```

Comment:
this does not seem to be defined, right? I will defined it or something like it.

Reply:
"SET" and "CLEAR" are defined for all 1 bit fields

Reply:
the name should not be set and clear anyway, more like BaudDivisor and DataMode or some such?

Reply:
Since DLAB is Divisor Latch Access Bit, you would want to name it in a way that signifies that access to the divisor latch is being enabled or disabled. In which case Enable and Disable are more fitting, assuming the reader knows what DLAB is.

Reply:
e.g.:
let control: InMemoryRegister<u32, Control::Register> = InMemoryRegister::new(0)
control.write(Contol::BYTE_COUNT.val(0) +
              Contol::ENABLE::Yes +
              Contol::LENGTH.val(10));

Reply:
I'd forgotten about that. Don't we want to use the 'InMemoryRegister' type as well?

Reply:
I think the RegisterBlock gets us that?